tests/integration/lineage.js

```diff
@@ -0,0 +1,489 @@
+      expect(actual).excluding('_rev').to.deep.equal(no_contact);
+    });
+  });
+
+  it('fetchHydratedDoc attaches the full lineage for reports', function() {
+    return lineage.fetchHydratedDoc(report._id).then(actual => {
+      expect(actual.form).to.equal('A');
+      // parents
+      expect(actual.contact.name).to.equal(report_contact.name);
+      expect(actual.contact.parent.name).to.equal(report_parent.name);
+      expect(actual.contact.parent.parent.name).to.equal(report_grandparent.name);
+      expect(actual.parent).to.not.exist; // jshint ignore:line
+      // contacts of parents
+      expect(actual.contact.parent.contact.phone).to.equal('+123');
+      expect(actual.contact.parent.parent.contact.phone).to.equal('+456');
+      // patient
+      expect(actual.patient.name).to.equal(report_patient.name);
+      // patient parents
+      expect(actual.patient.parent.name).to.equal(report_contact.name);
+    });
+  });
+
+  it('fetchHydratedDoc attaches the contacts', function() {
+    return lineage.fetchHydratedDoc(place._id).then(actual => {
+      expect(actual.name).to.equal(place.name);
+      expect(actual.contact.phone).to.equal('+789');
+      expect(actual.parent.name).to.equal(place_parent.name);
+      expect(actual.parent.contact.phone).to.equal('+123');
+      expect(actual.parent.parent.name).to.equal(place_grandparent.name);
+      expect(actual.parent.parent.contact.phone).to.equal('+456');
+    });
+  });
+
+  // This is a classic use-case: report from CHW who is the contact for their own area
+  it('fetchHydratedDoc attaches re-used contacts, minify handles the circular references', function() {
+    return lineage.fetchHydratedDoc(circular_report._id).then(actual => {
+      // The contact and the contact's parent's contact are the hydrated CHW
+      expect(actual.contact.parent.contact.hydrated).to.equal(true);
+      expect(actual.contact.hydrated).to.equal(true);
+
+      // And we can minifiy back to the original without error
+      lineage.minify(actual);
+      expect(actual).excludingEvery('_rev').to.deep.equal(circular_report);
+    });
+  });
+
+  it('fetchHydratedDoc fetches contacts that it has not got via lineage', function() {
+    sinon.spy(db, 'allDocs');
+    return lineage.fetchHydratedDoc(fetch_contact._id).then(() => {
+      expect(db.allDocs.getCall(0).args[0]).to.deep.equal({ keys: [dummyDoc2._id], include_docs: true });
+    });
+  });
+
+  it('fetchHydratedDoc does not fetch contacts that it has already got via lineage', function() {
+    sinon.spy(db, 'allDocs');
+    return lineage.fetchHydratedDoc(fetch_noContact._id).then(() => {
+      expect(db.allDocs.callCount).to.equal(0);
+    });
+  });
+
+  it('fetchHydratedDoc+minify is noop on a report', function() {
+    return lineage.fetchHydratedDoc(report._id).then(actual => {
+      lineage.minify(actual);
+      expect(actual).excluding('_rev').to.deep.equal(report);
```

Comment:
This test looks like it assumes that `fetchHydratedDoc()` returns a new object.  If it modifies the `report` in-place, then I suspect this test will pass even if `fetchHydratedDoc+minify` is _not_ a no-op.

Reply:
Good point, I'll do some cloning